### PR TITLE
Ravil/dr gpu

### PR DESCRIPTION
**Pull request by ravil-mobile**
This PR contains GPU implementation of space time interpolation which is required by DR cells. Additionally, pipelining has been utilized to hide CPU-to-GPU memory transfers which happens due to the DR part is still computed with CPUs. A DR layer is split into batches which forms a work-load for each stage. The first stage moves asynchronously data from device to host, the second stage runs a friction solver, the third stage asynchronously copies results computed by the second stage in its previous iteration. Because it is difficult to find a good batch size, a simple tuner has been implemented which takes into account estimated performance of the second stage.  (The DR pipeline is meant to be deprecated once DR is fully adapted for GPU.)

Pipelining was implemented in a generic way which allows to customize it for different kinds of computations if it is needed. Tests are included.

Additionally, changes were tested with 2 GPUs on an LRZ ARM+NVIDIA node. Numerical results are correct

### CODE REVIEW

File: auto_tuning/proxy/src/proxy_seissol_allocator.hpp [hunk 1]
```diff
@@ -248,13 +250,37 @@ void initDataStructuresOnDevice() {
...
+    constexpr size_t QInterpolatedSize = CONVERGENCE_ORDER * tensor::QInterpolated::size() * sizeof(real);
+    constexpr size_t imposedStateSize = tensor::QInterpolated::size() * sizeof(real);
+    constexpr size_t idofsSize = tensor::Q::size() * sizeof(real);
+
+    drLayer.setScratchpadSize(m_dynRup.QInterpolatedPlusOnDevice, QInterpolatedSize * drLayerSize);
+    drLayer.setScratchpadSize(m_dynRup.QInterpolatedMinusOnDevice, QInterpolatedSize * drLayerSize);
+    drLayer.setScratchpadSize(m_dynRup.idofsPlusOnDevice, idofsSize * drLayerSize);
+    drLayer.setScratchpadSize(m_dynRup.idofsMinusOnDevice, idofsSize * drLayerSize);
+
+    constexpr auto UpperStageFactor = dr::pipeline::DrPipeline::TailSize * dr::pipeline::DrPipeline::DefaultBatchSize;
```
Comment (krenzland) [inline auto_tuning/proxy/src/proxy_seissol_allocator.hpp:273]: Why is this in upper case?

File: src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    std::vector<real *> idofsPlusPtrs{};
+    std::vector<real *> idofsMinusPtrs{};
+
+    const size_t idofsSize = tensor::Q::size();
+    for (unsigned face = 0; face < currentLayer->getNumberOfCells(); ++face) {
+      timeDerivativePlusPtrs.push_back(timeDerivativePlus[face]);
+      timeDerivativeMinusPtrs.push_back(timeDerivativeMinus[face]);
+      idofsPlusPtrs.push_back(&idofsPlus[face * idofsSize]);
+      idofsMinusPtrs.push_back(&idofsMinus[face * idofsSize]);
+    }
```
Comment (krenzland) [inline src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp:35]: Relevant only if this thing here is called often (and not e.g. in init):
You could use reserve(size) on the vector or allocate it at once. Would save some re-allocations. If it isn't in a relevant loop, ignore this :)

**Review by krenzland — CHANGES_REQUESTED** [review]
Looks generally good, but I'm not sure I've understood everything correctly. See some attached comments. With which scenario did you test this?

P.S: Your pipeline has graduated so quickly! (I think I missed the Pipeline, M.Sc phase)

File: auto_tuning/proxy/src/proxy_seissol_allocator.hpp (hunk 1 above)
Comment (ravil-mobile) [reply auto_tuning/proxy/src/proxy_seissol_allocator.hpp:273]: `UpperStageFactor` is in upper case because it is a `constexpr`. I thought we have a convention that all constants should start with upper case.

DrPipeline is derived from GenericPipeline pipeline and it has only 3 stages. 

I guess the following figure make the terminology clear:

pipeline

File: src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp (hunk 2 above)
Comment (ravil-mobile) [reply src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp:35]: Yes, these loop are getting executed only once in the initialization stage. I agree about this particular loop. However, there are some places where I don't know sizes in advance.

File: src/Initializer/BatchRecorders/NeighIntegrationRecorder.cpp [hunk 3]
```diff
@@ -108,16 +104,15 @@ void NeighIntegrationRecorder::recordNeighbourFluxIntegrals() {
   for (unsigned cell = 0; cell < currentLayer->getNumberOfCells(); ++cell) {
     auto data = currentLoader->entry(cell);
     for (unsigned int face = 0; face < 4; face++) {
+      switch (data.cellInformation.faceTypes[face]) {
+        case FaceType::regular:
+          // Fallthrough intended
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Initializer/BatchRecorders/NeighIntegrationRecorder.cpp:109]: Thanks for the link. I will do it.

File: src/Initializer/BatchRecorders/Recorders.h [hunk 4]
```diff
@@ -69,17 +65,19 @@ class CompositeRecorder : public AbstractRecorder {
...
 
-class LocalIntegrationRecorder : public AbstractRecorder {
+class LocalIntegrationRecorder : public AbstractRecorder<seissol::initializers::LTS> {
 public:
   void record(LTS &handler, Layer &layer) override;
 
 private:
   void setUpContext(LTS &handler, Layer &layer, kernels::LocalData::Loader &loader) {
     currentLoader = &loader;
+    idofsAddressCounter = 0;
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Initializer/BatchRecorders/Recorders.h:79]: Ok

File: src/Solver/Pipeline/DrPipeline.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
+/**
+ * @file
+ * This file is part of SeisSol.
+ *
+ * @author Ravil Dorozhinskii (ravil.dorozhinskii AT tum.de)
+ *
+ * @section LICENSE
+ * Copyright (c) 2015-2017, SeisSol Group
+ * All rights reserved.
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Solver/Pipeline/DrPipeline.h:9]: Sure, will be fixed

File: src/Solver/Pipeline/DrTuner.cpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+    const auto diff = std::abs(leftPoint - rightPoint);
+    if (eps > diff) {
+      // convergence achieved
+      action = Action::SkipAction;
+      isConverged = true;
+      return;
+    }
+
+    // Golden-section search
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Solver/Pipeline/DrTuner.cpp:105]: Yep. will be done

File: src/Solver/Pipeline/GenericPipeline.h [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+    };
+
+    GenericPipeline(bool resetAfterRun = true) : resetAfterRun(resetAfterRun) {
+      using BaseTunerT = PipelineTuner<NumStagesP, DefaultBatchSizeP>;
+      static_assert(std::is_base_of_v<BaseTunerT, TunerT>,
+                    "ConcreteTunerT must be derived from PipelineTuner");
+    }
+    ~GenericPipeline() = default;
+    constexpr static decltype(NumStagesP) NumStages{NumStagesP};
+    static constexpr decltype(NumStagesP) TailSize{NumStagesP - 1};
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Solver/Pipeline/GenericPipeline.h:100]: Yep

File: src/tests/Pipeline/PipelineTest.t.h [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+  }
+
+public:
+  void testGeneral() {
+    TestPipeline pipeline{};
+    TS_ASSERT_EQUALS(TestPipeline::NumStages, 4);
+    TS_ASSERT_EQUALS(TestPipeline::TailSize, 3);
+  }
+
+  void testSuperShortPipeline()
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/tests/Pipeline/PipelineTest.t.h:74]: My fault. will be fixed

File: src/Solver/Pipeline/GenericPipeline.h [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+    size_t numTotalIterations{0};
+    size_t numFullPipeIterations{0};
+    std::array<Range, NumStages> ranges;
+    std::array<double, NumStages> stageTiming{};
+    std::array<PipelineCallBack*, NumStages> callBacks{nullptr};
+    bool resetAfterRun{true};
+    TunerT tuner;
+    Stopwatch stopwatch{};
+
+    friend class seissol::unit_test::PipelineTest;
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Solver/Pipeline/GenericPipeline.h:229]: Ok. There are two opinions about friend classes for unit testing. The ones who say "don't do it" mean that a class implementation may change in the future which will definitely lead to some changes in tests. I agree that friends create a tight coupling between classes and their tests. The ones who say "it is ok" mean that they want to avoid writing lot's of getter methods that nobody will use except test classes. In my case, a client only need two public methods, namely: `run` and `registerCallBack`.

I am personally in between of these two opinions because I agree with both of them. There is no strict preference of one approach over another one in the discussion that you attached. People say that sometimes it is ok to have friends for testing. 

In short, I need to access 2 private attributes of a pipeline: `numFullPipeIterations` and `numTotalIterations`. I can write getters for them.

What do you think?

File: src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+    if (!idofsPlusPtrs.empty()) {
+      ConditionalKey key(*KernelNames::DrTime);
+      checkKey(key);
+
+      (*currentTable)[key].content[*EntityId::DrDerivativesPlus] = new BatchPointers(timeDerivativePlusPtrs);
+      (*currentTable)[key].content[*EntityId::DrDerivativesMinus] = new BatchPointers(timeDerivativeMinusPtrs);
+      (*currentTable)[key].content[*EntityId::DrIdofsPlus] = new BatchPointers(idofsPlusPtrs);
+      (*currentTable)[key].content[*EntityId::DrIdofsMinus] = new BatchPointers(idofsMinusPtrs);
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp:44]: So, `currentTable` is a member of `LTStree`. Each layer has its own conditional table 
```
  // a part of class AbstractRecorder
  void setUpContext(LtsT &handler, Layer &layer) {
    currentTable = &(layer.getCondBatchTable());
```
Here is a definition:
```
using ConditionalBatchTableT =
    std::unordered_map<ConditionalKey, BatchTable, ConditionalHash<ConditionalKey>>;
}
```

In its turn, BatchTable is a wrapper of std::array which is responsible for deleting `BatchPointers`
```
struct BatchTable {
public:
  BatchTable() {
    for (auto &item : content) {
      item = nullptr;
    }
  }
  ~BatchTable() {
    for (auto &item : content) {
      delete item;
    }
  }

  std::array<BatchPointers *, *EntityId::Count> content{};
};
```

Here is a constructor and destructor of BatchPointers:
```
  explicit BatchPointers(std::vector<real *> collectedPointers)
      : pointers(std::move(collectedPointers)), devicePtrs(nullptr) {
    if (!pointers.empty()) {
      devicePtrs = (real **)device.api->allocGlobMem(pointers.size() * sizeof(real *));
      device.api->copyTo(devicePtrs, pointers.data(), pointers.size() * sizeof(real *));
    }
  }

  virtual ~BatchPointers() {
    if (devicePtrs != nullptr) {
      device.api->freeMem(devicePtrs);
      devicePtrs = nullptr;
    }
  }
```
LTSTree belongs to MemoryManager and it belongs to SeisSol class. Currently MemoryManager is a unique pointer in SeisSol class
```
  std::unique_ptr<initializers::MemoryManager> m_memoryManager{nullptr};
```

There is a method which we call from fortran at the very of an execution
```
void deleteMemoryManager() {
  m_memoryManager.reset(nullptr);
}
```
which is supposed to call m_ memoryManager destructor.

So, in short, according to me, everything should be deleted

[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp:44]: yes, I record raw pointers to GPU memory which will later be passed to yateto/gemmforge generated GPU kernels

File: auto_tuning/proxy/src/proxy_seissol_allocator.hpp (hunk 1 above)
Comment (krenzland) [reply auto_tuning/proxy/src/proxy_seissol_allocator.hpp:273]: Hm, I would introduce upper case for constants (except maybe static constants) just because I try to declare everything const that can be const.

File: src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp (hunk 2 above)
Comment (krenzland) [reply src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp]: Okay, if they are only in the init that's fine.

File: src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (krenzland) [reply src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp]: Thanks for the very detailed explanation! Are you sure that the destructors of the BatchPointers is actually called? I don't think so, unless I'm missing something. Could you maybe check? This is not an huge issue but it makes debugging with e.g. valgrind very hard because you get some strange output at the end of the program.

File: src/Solver/Pipeline/GenericPipeline.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (krenzland) [reply src/Solver/Pipeline/GenericPipeline.h]: Imo unit tests should only test the public interface

File: src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp]: I didn’t use valgrid because there is always some clutter caused by MPI. I know that there are some filter files for OpenMPI but it didn’t work for me. Tell me if you use valgrind for SeisSol.

When an LTS layer dies and destroys its instance of ConditionalBatchTableT (which is unordered_map) using its destructor. Each STL container calls each of the contained element's destructors while destructing itself. Therefore, destructor of BatchTable instances are going to be called which, in turn, are going to delete all BatchPointers.

File: src/Solver/Pipeline/GenericPipeline.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Solver/Pipeline/GenericPipeline.h]: Done, Friend classes have already been removed

File: src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp]: I added a print statement to the BatchPointers destructor to just as a sanity check and ran SeisSol. At the end I got a bunch of prints on my console. So, it wokrs

[the remark this answers is not in the document]
Comment (krenzland) [reply src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp]: Ah, sorry. I misunderstood your explanation. Read it again, and it's clear.

**Review by krenzland — APPROVED** [review]
